kotest-assertions/kotest-assertions-json/src/jvmMain/kotlin/io/kotest/assertions/json/JsonMatchers.kt

```diff
@@ -1,81 +1,201 @@
 package io.kotest.assertions.json
 
-import com.fasterxml.jackson.databind.ObjectMapper
-import com.fasterxml.jackson.module.kotlin.registerKotlinModule
 import com.jayway.jsonpath.JsonPath
 import com.jayway.jsonpath.PathNotFoundException
+import io.kotest.assertions.failure
 import io.kotest.matchers.Matcher
 import io.kotest.matchers.MatcherResult
 import io.kotest.matchers.should
 import io.kotest.matchers.shouldNot
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.contract
+import kotlin.reflect.typeOf
+
+@OptIn(ExperimentalContracts::class)
+infix fun Json?.shouldMatchJson(json: Json?) {
+    contract {
+        returns() implies (this@shouldMatchJson != null)
+    }
+
+    this should matchJson(json)
+}
+
+infix fun Json?.shouldNotMatchJson(json: Json?) = this shouldNot matchJson(json)
+
+fun matchJson(json: Json?) = object : Matcher<Json?> {
+
+    override fun test(value: Json?): MatcherResult {
+        val actualJson = value?.let { publishedMapper.readTree(it) }
+        val expectedJson = json?.let { publishedMapper.readTree(it) }
 
-private val mapper by lazy { ObjectMapper().registerKotlinModule() }
+        return MatcherResult(
+            actualJson == expectedJson,
+            "expected: $expectedJson but was: $actualJson",
+            "expected not to match with: $expectedJson but match: $actualJson"
+        )
+    }
+}
 
-infix fun String.shouldMatchJson(json: String) = this should matchJson(json)
-infix fun String.shouldNotMatchJson(json: String) = this shouldNot matchJson(json)
-fun matchJson(json: String) = object : Matcher<String> {
+infix fun Json.shouldMatchJsonResource(resource: String) = this should matchJsonResource(resource)
+infix fun Json.shouldNotMatchJsonResource(resource: String) = this shouldNot matchJsonResource(resource)
 
-  override fun test(value: String): MatcherResult {
+fun matchJsonResource(resource: String) = object : Matcher<Json> {
```

Comment:
Thanks for making JSON matchers work on nullable values. I just reported issue with it (https://github.com/kotest/kotest/issues/1504) and now I noticed you are addressing it in your PR. I have noticed however, that in your changes some matchers should work with nullable values (`shouldMatchJson`, `shouldContainJsonKey`), but some don't (for example `shouldMatchJsonResource` and `shouldNotContainJsonKey` will not work on nullable strings). Could you change it to make all those matchers consistently work with nullable values?

Reply:
I just don't change this `matchJsonResource` matcher :) But thank you for noticing, if there is demand for it, I'll do it too.

Will continue work on this PR in about a week, planning to finish to the end of June.

Reply:
That would be great :) Sorry to piggyback to your PR, but I figured it will be easier for everyone this way, instead of doing separate small PR and having to resolve merge conflicts afterwards.

Reply:
Hey @klesniewski , I've added nullable receiver for `shouldMatchJsonResource`. Is it enough?

I'm not adding nullable receiver support for the most of negated versions as it's not clear how to treat it. So I'm not adding nullable receiver for `shouldNotMatchJsonResource` now.